File: 735-asteroid-collision/asteroid-collision.ts
function asteroidCollision(asteroids: number[]): number[] {
  const stack: number[] = [];

  for (const asteroid of asteroids) {
    if (asteroid > 0) {
      stack.push(asteroid);
    } else {
      while (stack.length > 0 && stack[stack.length - 1] > 0 && stack[stack.length - 1] < Math.abs(asteroid)) {
        stack.pop();
      }
      if (stack.length == 0 || stack[stack.length - 1] < 0) {
        stack.push(asteroid);
      } else if (stack[stack.length - 1] == Math.abs(asteroid)) {
        stack.pop();
      }
    }
  }
  return stack;
}

File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        s = []
        for e in asteroids:
            while s and s[-1] > 0 and e <0:
                is_col = e + s[-1]
                if is_col > 0 :
                    e = 0
                elif is_col < 0:
                    s.pop()
                else:
                    s.pop()
                    e=0
            if e:
                s.append(e)
        return s